algebra/src/algebras/tropical.rs

```diff
@@ -0,0 +1,252 @@
+//! ```
+//! use harness_algebra::algebras::tropical::TropicalAlgebra;
+//!
+//! let algebra = TropicalAlgebra::new();
+//! let a = algebra.element(3.0);
+//! let b = algebra.element(5.0);
+//!
+//! // Addition is max: 3 ⊕ 5 = 5
+//! let sum = a + b;
+//! assert_eq!(sum.value(), 5.0);
+//!
+//! // Multiplication is +: 3 ⊗ 5 = 8
+//! let product = a * b;
+//! assert_eq!(product.value(), 8.0);
+//! ```
+
+use std::fmt::{Debug, Display, Formatter};
+
+use crate::{
+  arithmetic::{Add, AddAssign, Additive, Mul, MulAssign, Multiplicative, One, Zero},
+  semiring::Semiring,
+};
+
+/// A tropical algebra element.
+///
+/// This represents a value in the tropical semiring (ℝ ∪ {-∞}, max, +).
+/// The value is stored as an `f64`, with `f64::NEG_INFINITY` representing -∞.
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+pub struct TropicalElement {
+  value: f64,
+}
+
+impl TropicalElement {
+  /// Creates a new tropical element with the given value.
+  pub fn new(value: f64) -> Self { Self { value } }
+
+  /// Returns the value of this tropical element.
+  pub fn value(&self) -> f64 { self.value }
+}
+
+impl Display for TropicalElement {
+  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result { write!(f, "{}", self.value) }
+}
+
+impl Add for TropicalElement {
+  type Output = Self;
+
+  fn add(self, rhs: Self) -> Self::Output { Self { value: self.value.max(rhs.value) } }
+}
+
+impl AddAssign for TropicalElement {
+  fn add_assign(&mut self, rhs: Self) { *self = *self + rhs; }
+}
+
+impl Mul for TropicalElement {
+  type Output = Self;
+
+  #[allow(clippy::suspicious_arithmetic_impl)]
+  fn mul(self, rhs: Self) -> Self::Output { Self { value: self.value + rhs.value } }
+}
+
+impl MulAssign for TropicalElement {
+  fn mul_assign(&mut self, rhs: Self) { *self = *self * rhs; }
+}
+
+impl Additive for TropicalElement {}
+impl Multiplicative for TropicalElement {}
+
+impl Zero for TropicalElement {
+  fn zero() -> Self { Self { value: f64::NEG_INFINITY } }
+
+  fn is_zero(&self) -> bool { self.value == f64::NEG_INFINITY }
+}
+
+impl One for TropicalElement {
+  fn one() -> Self { Self { value: 0.0 } }
+}
+
+impl Semiring for TropicalElement {
+  fn zero() -> Self { <Self as Zero>::zero() }
+
+  fn one() -> Self { <Self as One>::one() }
+}
+
+/// A tropical algebra.
+///
+/// This struct provides a way to create tropical elements.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct TropicalAlgebra;
+
+impl TropicalAlgebra {
+  /// Creates a new tropical algebra.
+  pub fn new() -> Self { Self }
+
+  /// Creates a new tropical element with the given value.
+  pub fn element(&self, value: f64) -> TropicalElement { TropicalElement::new(value) }
+
+  /// Returns the zero element of the tropical algebra (-∞).
+  pub fn zero_element(&self) -> TropicalElement { <TropicalElement as Zero>::zero() }
+
+  /// Returns the one element of the tropical algebra (0).
+  pub fn one_element(&self) -> TropicalElement { <TropicalElement as One>::one() }
+}
```

Comment:
This doesn't feel necessary here. It was necessary in the `CliffordAlgebra` example because the `CliffordAlgebra` itself has to hold on to more data -- the `QuadraticForm`. `CliffordAlgebraElements` then own a reference to the `&QuadraticForm`. So this could be consolidated.